Take input striding for conv forward based on eager output

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88891
* __->__ #88706
* #88534



cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

From discussion with @Chillee and @ngimel we'll likely need further fixes to ensure that we hit channels last kernels but this is still worth landing in its own right.

eellison committed:
Take input striding for conv based on eager output

[ghstack-poisoned]

ngimel commented:
Can you also make FixedLayout instead of FlexibleLayout? That'll make cl propagation more robust. 
A time will come (soon) when we'll want to use cl in inductor even when eager was channels first, so we'll have to rethink the strategy of taking strides from eager, but for now to enable cl perf that's fine.

eellison committed:
Update on "Take input striding for conv forward based on eager output"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Take input striding for conv forward based on eager output"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper commented on torch/_inductor/ir.py:
```diff
@@ -3016,9 +3016,27 @@ def create(
+            bias_fake = (
+                ir_node_to_tensor(bias, guard_shape=False) if bias is not None else bias
+            )
+            output = torch.ops.aten.convolution(
```
If we add this path to query output strider order, do we still need the following path: 

````
  with FakeTensorMode():
            output, *_ = cls.process_kernel(
                torch.ops.aten.convolution,
                x,
                weight,
                bias,
                stride,
                padding,
                dilation,
                transposed,
                output_padding,
                groups,
            )
```

eellison committed:
Update on "Take input striding for conv forward based on eager output"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Take input striding for conv forward based on eager output"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ngimel reviewed (approved):
The test is weak (it would have passed even without this PR), but that's fine.

Chillee commented:
Don't we also need to do this for convolutionbackward?

eellison commented:
> The test is weak (it would have passed even without this PR), but that's fine.

Not saying it's strong but prior we were taking striding for the weight

ngimel commented:
Hm, before this PR cl input + contig weight did cl convolution

eellison committed:
Update on "Take input striding for conv forward based on eager output"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

From discussion with Chillee and ngimel we'll likely need further fixes to ensure that we hit channels last kernels but this is still worth landing in its own right.

[ghstack-poisoned]

eellison committed:
Update on "Take input striding for conv forward based on eager output"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

From discussion with Chillee and ngimel we'll likely need further fixes to ensure that we hit channels last kernels but this is still worth landing in its own right.

[ghstack-poisoned]

jgong5 commented on torch/_inductor/ir.py:
```diff
@@ -3016,32 +3024,39 @@ def create(
+            )
+            req_stride_order = get_stride_order(output.stride())
+
+        if config.triton.convolution == "aten":
+            weight = cls.require_stride_order(weight, req_stride_order)
+            x = cls.require_stride_order(x, req_stride_order)
```
For CPU convolution, we would always convert the weight to the blocked layout inside the ATen conv. If the stride order of `weight` is different from output, there would be extra copy by requiring `weight` to follow the output stride. Shall we keep the weight stride intact for CPU backend here?
```suggestion
        if is_cpu():
            weight = cls.require_stride1(cls.realize_input(weight))
            x = cls.require_stride_order(x, req_stride_order)        
        elif config.triton.convolution == "aten":
            weight = cls.require_stride_order(weight, req_stride_order)
            x = cls.require_stride_order(x, req_stride_order)
```

eellison commented:
@pytorchbot merge
